# Pandas — Series, DataFrame, eksploracja danych

## Ćwiczenie 1: Series — tworzenie i operacje

In [3]:
import pandas as pd
import numpy as np

# Series z listy — domyślny indeks (0, 1, 2, ...)
oceny = pd.Series([4.5, 3.0, 5.0, 3.5, 4.0])
print(oceny)

0    4.5
1    3.0
2    5.0
3    3.5
4    4.0
dtype: float64


Zadanie 1a: Utwórz Series pensje z dict:

In [4]:
pensje = pd.Series({
    'Anna': 5500,
    'Jan': 7200,
    'Ewa': 4800,
    'Marek': 9100,
    'Kasia': 6300
})

print(pensje)

Anna     5500
Jan      7200
Ewa      4800
Marek    9100
Kasia    6300
dtype: int64


Zadanie 1b: Odpowiedz na pytania o pensje

1. Wyświetl imię osoby zarabiającej najwięcej (użyj idxmax())


In [4]:
najlepszy = pensje.idxmax()
print("Najwięcej zarabia:", najlepszy)

Najwięcej zarabia: Marek


2. Oblicz średnią pensję

In [5]:
srednia = pensje.mean()
print("Średnia pensja:", srednia)

Średnia pensja: 6580.0


3. Wyświetl imiona osób zarabiających powyżej średniej (filtrowanie boolean)

In [6]:
powyzej = pensje[pensje > srednia]

print("Powyżej średniej:")
print(powyzej)

Powyżej średniej:
Jan      7200
Marek    9100
dtype: int64


4. Policz ile osób zarabia poniżej 6000 (użyj len() lub .sum())

In [ ]:
ponizej_6000 = (pensje < 6000).sum()

print("Ile osób < 6000:", ponizej_6000)

Ile osób < 6000: 2


Zadanie 1c: Oblicz pensje po zmianach:

1. Oblicz pensje po podwyżce 10% dla wszystkich (pensje * 1.10)

In [5]:
podwyzka = pensje * 1.10
print("Po podwyżce:")
print(podwyzka)

Po podwyżce:
Anna      6050.0
Jan       7920.0
Ewa       5280.0
Marek    10010.0
Kasia     6930.0
dtype: float64


2. Oblicz pensje po premii 500 zł dla wszystkich (pensje + 500)

In [6]:
premia = pensje + 500
print("Po premii:")
print(premia)

Po premii:
Anna     6000
Jan      7700
Ewa      5300
Marek    9600
Kasia    6800
dtype: int64


3. Oblicz nową pensję = (pensja * 1.10) + 500

In [7]:
nowe_pensje = (pensje * 1.10) + 500

print("Nowe pensje:")
print(nowe_pensje)

Nowe pensje:
Anna      6550.0
Jan       8420.0
Ewa       5780.0
Marek    10510.0
Kasia     7430.0
dtype: float64


4. Wyświetl ranking nowych pensji od najwyższej (użyj sort_values)

In [8]:
ranking = nowe_pensje.sort_values(ascending=False)

print("Ranking nowych pensji:")
print(ranking)

Ranking nowych pensji:
Marek    10510.0
Jan       8420.0
Kasia     7430.0
Anna      6550.0
Ewa       5780.0
dtype: float64


## Ćwiczenie 2: DataFrame — tworzenie i atrybuty

In [9]:
# Dane o produktach sklepu internetowego
sklep = pd.DataFrame({
    'produkt': ['Laptop', 'Tablet', 'Smartfon', 'Słuchawki', 'Monitor'],
    'cena': [3500, 1800, 2500, 350, 2200],
    'stan_magazynowy': [45, 120, 200, 500, 75],
    'kategoria': ['Komputery', 'Mobilne', 'Mobilne', 'Akcesoria', 'Komputery']
})

### Zadanie 2a: Zbadaj DataFrame sklep

1. Wyświetl kształt tabeli (shape)

In [10]:
print("Shape:", sklep.shape)

Shape: (5, 4)


2. Wyświetl nazwy kolumn (columns)

In [11]:
print("Kolumny:", sklep.columns)

Kolumny: Index(['produkt', 'cena', 'stan_magazynowy', 'kategoria'], dtype='str')


3. Wyświetl typy danych w każdej kolumnie (dtypes)

In [12]:
print("Typy danych:")
print(sklep.dtypes)

Typy danych:
produkt              str
cena               int64
stan_magazynowy    int64
kategoria            str
dtype: object


4. Wyświetl pierwsze 3 wiersze (head)

In [13]:
print(sklep.head(3))

    produkt  cena  stan_magazynowy  kategoria
0    Laptop  3500               45  Komputery
1    Tablet  1800              120    Mobilne
2  Smartfon  2500              200    Mobilne


### Zadanie 2b: Dodaj nowe kolumny do DataFrame sklep

1. Dodaj kolumnę 'wartosc_magazynu' = cena * stan_magazynowy

In [14]:
sklep['wartosc_magazynu'] = sklep['cena'] * sklep['stan_magazynowy']

2. Dodaj kolumnę 'cena_z_vat' = cena * 1.23

In [15]:
sklep['cena_z_vat'] = sklep['cena'] * 1.23

3. Wyświetl cały DataFrame z nowymi kolumnami

In [16]:
print(sklep)

     produkt  cena  stan_magazynowy  kategoria  wartosc_magazynu  cena_z_vat
0     Laptop  3500               45  Komputery            157500      4305.0
1     Tablet  1800              120    Mobilne            216000      2214.0
2   Smartfon  2500              200    Mobilne            500000      3075.0
3  Słuchawki   350              500  Akcesoria            175000       430.5
4    Monitor  2200               75  Komputery            165000      2706.0


In [17]:
# Wczytaj dataset tips
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv'
tips = pd.read_csv(url)

### Zadanie 2c: Podstawowa eksploracja datasetu tips

1. Wyświetl ile wierszy i kolumn ma dataset (shape)

In [18]:
print("Shape:", tips.shape)

Shape: (244, 7)


2. Wyświetl ostatnie 5 wierszy (tail)

In [19]:
print(tips.tail())

     total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2


3. Wyświetl 3 losowe wiersze (sample)

In [20]:
print(tips.sample(3))

     total_bill   tip     sex smoker  day    time  size
174       16.82  4.00    Male    Yes  Sun  Dinner     2
1         10.34  1.66    Male     No  Sun  Dinner     3
94        22.75  3.25  Female     No  Fri  Dinner     2


4. Wyświetl typy danych w każdej kolumnie (dtypes)

In [21]:
print(tips.dtypes)

total_bill    float64
tip           float64
sex               str
smoker            str
day               str
time              str
size            int64
dtype: object


## Ćwiczenie 3: Eksploracja datasetu

In [23]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv'
penguins = pd.read_csv(url)

### Zadanie 3a: Pierwszy rzut oka 

1. Wyświetl head(10)

In [24]:
print(penguins.head(10))

  species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       36

2. Wyświetl shape

In [25]:
print("Shape:", penguins.shape)

Shape: (344, 7)


3. Uruchom info() — ile brakujących wartości widzisz?

In [26]:
penguins.info()

<class 'pandas.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    str    
 1   island             344 non-null    str    
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    str    
dtypes: float64(4), str(3)
memory usage: 18.9 KB


### Zadanie 3b: Statystyki opisowe

1. Uruchom describe() — co mówią ci średnie i odchylenia?

In [27]:
print(penguins.describe())

       bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      342.000000     342.000000         342.000000   342.000000
mean        43.921930      17.151170         200.915205  4201.754386
std          5.459584       1.974793          14.061714   801.954536
min         32.100000      13.100000         172.000000  2700.000000
25%         39.225000      15.600000         190.000000  3550.000000
50%         44.450000      17.300000         197.000000  4050.000000
75%         48.500000      18.700000         213.000000  4750.000000
max         59.600000      21.500000         231.000000  6300.000000


2. Uruchom describe(include='all') — co dodają kolumny tekstowe?

In [28]:
print(penguins.describe(include='all'))

       species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
count      344     344      342.000000     342.000000         342.000000   
unique       3       3             NaN            NaN                NaN   
top     Adelie  Biscoe             NaN            NaN                NaN   
freq       152     168             NaN            NaN                NaN   
mean       NaN     NaN       43.921930      17.151170         200.915205   
std        NaN     NaN        5.459584       1.974793          14.061714   
min        NaN     NaN       32.100000      13.100000         172.000000   
25%        NaN     NaN       39.225000      15.600000         190.000000   
50%        NaN     NaN       44.450000      17.300000         197.000000   
75%        NaN     NaN       48.500000      18.700000         213.000000   
max        NaN     NaN       59.600000      21.500000         231.000000   

        body_mass_g   sex  
count    342.000000   333  
unique          NaN     2  
top

3. Jaka jest mediana masy ciała (body_mass_g)?

In [29]:
print("Mediana body_mass_g:", penguins['body_mass_g'].median())

Mediana body_mass_g: 4050.0


### Zadanie 3c: Rozkłady wartości

1. Ile jest pingwinów każdego gatunku? (value_counts na 'species')

In [30]:
print(penguins['species'].value_counts())

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64


2. Ile pingwinów na każdej wyspie? (value_counts na 'island')

In [31]:
print(penguins['island'].value_counts())

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64


3. Rozkład płci (value_counts na 'sex') — czy jest zbalansowany?

In [32]:
print(penguins['sex'].value_counts())

sex
MALE      168
FEMALE    165
Name: count, dtype: int64


4. Jaki procent to gatunek Adelie? (normalize=True)

In [33]:
print(penguins['species'].value_counts(normalize=True) * 100)

species
Adelie       44.186047
Gentoo       36.046512
Chinstrap    19.767442
Name: proportion, dtype: float64


### Zadanie 3d: Brakujące dane

1. Policz brakujące wartości w każdej kolumnie (isna().sum())

In [34]:
print(penguins.isna().sum())

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64


2. Jaki procent danych brakuje w kolumnie 'sex'?

In [35]:
braki = penguins['sex'].isna().sum()
procent = (braki / len(penguins)) * 100

print("Braki w sex (%):", procent)

Braki w sex (%): 3.1976744186046515


### Zadanie 3e: Wnioski

Wnioski z analizy datasetu penguins
- Dataset zawiera 344 pingwiny z 3 gatunków.
- Najczęstszy gatunek to Adelie (największa liczba obserwacji).
- Pingwiny występują na 3 wyspach: Biscoe, Dream i Torgersen.
- Średnia masa ciała wynosi ok. 4200 g (mediana ~ 4050 g).
- W kolumnie 'sex' występują braki danych.
- Największa populacja znajduje się na wyspie Biscoe.

## Ćwiczenie 4: Pytania biznesowe

In [36]:
tips = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv')

# Dodaj kolumnę z procentem napiwku
tips['tip_pct'] = (tips['tip'] / tips['total_bill'] * 100).round(1)

1. Oblicz średni rachunek (total_bill)

In [37]:
print("Średni rachunek:", tips['total_bill'].mean())

Średni rachunek: 19.78594262295082


2. Oblicz średni napiwek w procentach (tip_pct)

In [38]:
print("Średni tip %:", tips['tip_pct'].mean())

Średni tip %: 16.08155737704918


3. Wyświetl który dzień ma najwięcej rachunków (value_counts na 'day')

In [39]:
print(tips['day'].value_counts())

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64


4. Policz ile rachunków jest od palaczy (value_counts na 'smoker')

In [40]:
print(tips['smoker'].value_counts())

smoker
No     151
Yes     93
Name: count, dtype: int64


5. Wyświetl describe() tylko dla kolumny 'tip' — jaka jest mediana napiwku?

In [41]:
print(tips['tip'].describe())
print("Mediana tip:", tips['tip'].median())

count    244.000000
mean       2.998279
std        1.383638
min        1.000000
25%        2.000000
50%        2.900000
75%        3.562500
max       10.000000
Name: tip, dtype: float64
Mediana tip: 2.9


6. Jaki procent rachunków pochodzi z obiadu (Dinner)?

In [42]:
print(tips['time'].value_counts(normalize=True) * 100)

time
Dinner    72.131148
Lunch     27.868852
Name: proportion, dtype: float64


7. Ile unikalnych rozmiarów grup jest w danych? (nunique na 'size')

In [43]:
print("Ile różnych grup:", tips['size'].nunique())

Ile różnych grup: 6
